nixos/modules/services/networking/tinydns.nix

```diff
@@ -32,23 +220,48 @@ with lib;
   config = mkIf config.services.tinydns.enable {
     environment.systemPackages = [ pkgs.djbdns ];
 
-    users.users.tinydns = {};
+    users.users.tinydns = {
+      isSystemUser = true;
+      uid = config.ids.uids.tinydns;
+      group = "tinydns";
+    };
+    users.groups.tinydns = {
+      gid = config.ids.gids.tinydns;
+    };
 
     systemd.services.tinydns = {
       description = "djbdns tinydns server";
       wantedBy = [ "multi-user.target" ];
-      path = with pkgs; [ daemontools djbdns ];
+      environment = {
+        IP = config.services.tinydns.ip;
+        ROOT = "/run/tinydns";
+        UID = "${toString config.users.users.tinydns.uid}";
+        GID = "${toString config.users.groups.tinydns.gid}";
+      };
+      path = with pkgs; [ coreutils djbdns ];
+      serviceConfig = {
+        ExecStart = "${lib.getBin pkgs.djbdns}/bin/tinydns";
+        LimitDATA = "3000000";
+        RestartSec = 1;
+      };
       preStart = ''
-        rm -rf /var/lib/tinydns
-        tinydns-conf tinydns tinydns /var/lib/tinydns ${config.services.tinydns.ip}
-        cd /var/lib/tinydns/root/
-        ln -sf ${pkgs.writeText "tinydns-data" config.services.tinydns.data} data
+        rm -rf $ROOT
+        mkdir -m755 -p $ROOT
+        chown tinydns.tinydns $ROOT
+        cd $ROOT
+        ln -sf ${pkgs.writeText "tinydns-data" configTxt} data
         tinydns-data
       '';
-      script = ''
-        cd /var/lib/tinydns
-        exec ./run
+      reload = ''
+        cd $ROOT
+        ln -sf ${pkgs.writeText "tinydns-data" configTxt} data
```

Comment:
You're linking to a build time text file which has already been linked. For what purpose? This file won't change during the life of a nixos generation.

Reply:
The file `data` is merely the source input to the actual DNS database, which is constructed at the start of the service. The reason for this construction is threefold:
1. High availability. If for some reason, `nixos-rebuild` on the DNS server is not working but the database requires a change, the user can turn `data` into a separate copy, alter its content and update the database without disrupting DNS service.
2. Failover. If for some reason, the authoritative DNS server fails and a sub-ordinate DNS server (which would have received its database via some other means) is promoted, the user can turn `data` into a separate copy, etc.
3. To ensure continous service, the user must be able to recreate database while the service is running. The protocol of the database compiler mandates that the source file is in the same directory as the result, making it necessary to (at least) provide a link in the database's final location.

Reply:
Ah, yeah that all makes sense. Thanks for the answer.